### ISSUE TITLE
Can't use libnxz with OpenSSH

### ISSUE THREAD

**Issue by tuliom** [libnxz/power-gzip#74]
I tried to use libnxz with OpenSSH clients (ssh and scp), but the programs seem to be stuck in an infinite loop.

```
LD_PRELOAD=$(pwd)/lib/libnxz.so.0.62.2 scp -C oct/calgary.uncompressed server.example:/dev/null
```

I guess libnxz is not returning the same results as the executables are expecting.

**Comment by lucmaga** [comment]
It indeed stays on this loop from scp https://github.com/openssh/openssh-portable/blob/master/packet.c#L791-L819
```
	for (;;) {
		/* Set up fixed-size output buffer. */
		ssh->state->compression_in_stream.next_out = buf;
		ssh->state->compression_in_stream.avail_out = sizeof(buf);

		status = inflate(&ssh->state->compression_in_stream,
		    Z_PARTIAL_FLUSH);
		switch (status) {
		case Z_OK:
			if ((r = sshbuf_put(out, buf, sizeof(buf) -
			    ssh->state->compression_in_stream.avail_out)) != 0)
				return r;
			break;
		case Z_BUF_ERROR:
			/*
			 * Comments in zlib.h say that we should keep calling
			 * inflate() until we get an error.  This appears to
			 * be the error that we get.
			 */
			return 0;
		case Z_DATA_ERROR:
			return SSH_ERR_INVALID_FORMAT;
		case Z_MEM_ERROR:
			return SSH_ERR_ALLOC_FAIL;
		case Z_STREAM_ERROR:
		default:
			ssh->state->compression_in_failures++;
			return SSH_ERR_INTERNAL_ERROR;
		}
```

The scp code expects that after some time zlib returns `Z_BUF_ERROR`. At this point the stream have 0 bytes on `avail_in` but still have room on `avail_out`.
The zlib has a restriction that if no progress was made during a inflate call it will return `Z_BUF_ERROR`.  This can be seen on code at https://github.com/madler/zlib/blob/master/inflate.c#L1272. It is also on the inflate description at zlib.h https://github.com/madler/zlib/blob/master/zlib.h#L510-L512.

I'm trying to implement this behavior on libnxz. Any ideas?

**Comment by abalib** [comment]
Wild guess. This line needs more work https://github.com/libnxz/power-gzip/blob/master/lib/nx_inflate.c#L927 

https://github.com/madler/zlib/blob/master/inflate.c#L618
https://github.com/madler/zlib/blob/master/inflate.c#L1272
It looks like (if no input was read and no output was written OR if Z_FINISH) then change Z_OK to Z_BUF_ERROR

**Comment by abalib** [comment]
Also last sentence in this paragraph is interesting
https://github.com/madler/zlib/blob/master/inflate.c#L611  You asked me to Z_FINISH but you didn't give me enough output space or enough input to finish.  Therefore, I am returning Z_BUF_ERROR.  THat's how I read it.

**Comment by lucmaga** [comment]
@abalib I agree with your interpretation. But I understand we have two separate behaviors regarding `Z_BUF_ERROR`. The first one described by https://github.com/madler/zlib/blob/master/inflate.c#L611 as you pointed out. 
The second one is described in https://github.com/madler/zlib/blob/master/zlib.h#L510-L512 and implemented at https://github.com/madler/zlib/blob/master/inflate.c#L1272.
The line you pointed out is at the start of `nx_inflate_`. At this point we don't know if there we will have progress or not. I could guess that if `avail_in == 0` or `avail_out == 0` there is no room to progress. But maybe implementing it another way my cause some undesirable corner cases.

> Also last sentence in this paragraph is interesting
https://github.com/madler/zlib/blob/master/inflate.c#L611 You asked me to Z_FINISH but you didn't give me enough output space or enough input to finish. Therefore, I am returning Z_BUF_ERROR. THat's how I read it.

@abalib About this, notice that the `flush == Z_FINISH` in https://github.com/madler/zlib/blob/master/inflate.c#L1272 implements an other unrelated behavior. The issue here just because inflate didn't progress during the call as the scp just uses `Z_PARTIAL_FLUSH`.

**Comment by tuliom** [comment]
> About this, notice that the `flush == Z_FINISH` in https://github.com/madler/zlib/blob/master/inflate.c#L1272 implements an other unrelated behavior. The issue here just because inflate didn't progress during the call as the scp just uses `Z_PARTIAL_FLUSH`.

@lucmaga notice the code is using `||`, not `&&`.  That behavior also applies to `Z_PARTIAL_FLUSH`.

**Comment by tuliom** [comment]
I suspect the important part is this: https://github.com/madler/zlib/blob/master/inflate.c#L1248-L1250

**Comment by lucmaga** [comment]
> @lucmaga notice the code is using ||, not &&. That behavior also applies to Z_PARTIAL_FLUSH.
Thanks, I was not clear on that statement. I just wanted to say that the Z_FINISH condition is unrelated to the issue.
